### PR TITLE
[Core] Validate specification trait sets passed to `register`

### LINKED ISSUE #328: [Core] Validate constant trait sets in `register` batch
## What

Add validation in `Manager.register` that the supplied Specifications all have the same trait set.

## Why

The API contract requires that the specifications passed to `ManagerInterface.register` have constant trait sets (to allow the manager to optimise the batch submission). We should take advantage of the API middleware to validate this, so the `ManagerInterface` implementation doesn't need to check. 

This will add some overhead but:
 - We consider `register` to be a far lower call volume than `resolve` and so consider this overhead negligible for now.
 - Removing it later is safer than adding the check in later, once lax code is already in use.

### PR DESCRIPTION
**Pull request by foundrytom**
Closes #328.

This is a follow up PR to #327, so ignore anything other than [this commit](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/329/commits/a7f8ba6b78f453b68315e96f24e53d01cbf50415) until it is merged.

### CODE REVIEW

File: python/openassetio/hostAPI/Manager.py [hunk 1]
```diff
@@ -1077,6 +1077,15 @@ def register(self, targetEntityRefs, entitySpecs, context):
         if len(targetEntityRefs) != len(entitySpecs):
             raise IndexError("Parameter lists must be of the same length")
 
+        if entitySpecs:
+            # Check supplied specifcations share a trait set
+            expectedTraits = entitySpecs[0].traitIds()
```
Comment (foundrytom) [inline python/openassetio/hostAPI/Manager.py:1082]: [question] is `next(iter(entitySpecs))` preferred here, and manually tracking an index, on the off chance that this isn't a `list` or `tuple`?

Comment (feltech) [reply python/openassetio/hostAPI/Manager.py:1082]: An intriguing problem. A Pythonic one-liner, but doesn't short-circuit

```python
len(set(frozenset(s.traitIds()) for s in entitySpecs)) == 1
```

Ideally `traitIds()` would return a `frozenset` (needed to be used in a set of sets).

A quick google reveals solutions using `itertools.groupby`, though I suspect some `frozenset` shenanigans would be required for us, [e.g. from StackOverflow](https://stackoverflow.com/a/3844832/535103)

```python
from itertools import groupby

def all_equal(iterable):
    g = groupby(iterable)
    return next(g, True) and not next(g, False)
````

Comment (feltech) [reply python/openassetio/hostAPI/Manager.py:1082]: Ah but you're trying to make a nice error message, which limits the solution.

Comment (feltech) [reply python/openassetio/hostAPI/Manager.py:1082]: I think it's reasonable to assume, maybe even require, a sequential container that is indexable, so elements can correspond to elements in `targetEntityRefs`.  The only counterexample that springs to mind is a `dict` whose keys correspond, but that would require different treatment in the manager plugin, complicating matters.

**Review by feltech — APPROVED** [review]

**Comment by feltech** [comment]
LGTM, rebase and merge when ready.

**Pull request merged**